x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,260 @@
+			return nil, err
+		}
+		return nil, fmt.Errorf("username does not contain a valid prefix-suffix format")
+	}
+
+	usernamePrefix := parts[0] // Prefix (e.g., "jane")
+	usernameSuffix := parts[1] // Suffix (e.g., "-country-ir")
+
+	log.Printf("Attempting authentication with prefix: %s and suffix: %s", usernamePrefix, usernameSuffix)
+
+	// Validate the prefix and password against stored credentials
+	if !a.PrefixCredentials.Valid(usernamePrefix, password, userAddr) {
+		// Authentication failed
+		if _, err := writer.Write([]byte{statute.UserPassAuthVersion, statute.AuthFailure}); err != nil {
+			return nil, err
+		}
+		return nil, statute.ErrUserAuthFailed
+	}
+
+	// Authentication successful
+	if _, err := writer.Write([]byte{statute.UserPassAuthVersion, statute.AuthSuccess}); err != nil {
+		return nil, err
+	}
+
+	// Return the context with the extracted suffix for later use
+	return &socks5.AuthContext{
+		Method: statute.MethodUserPassAuth,
+		Payload: map[string]string{
+			"suffix": usernameSuffix, // Store the suffix for later use in the dialer
+			"prefix": usernamePrefix, // Store the prefix as well, if needed
+		},
+	}, nil
+}
+
+// CredentialStore is an interface to validate prefix-based credentials
+type CredentialStore interface {
+	Valid(username, password, userAddr string) bool
+}
+
+// StaticCredentialStore is a simple credential store that holds a map of valid prefixes and passwords
+type StaticCredentialStore struct {
+	credentials map[string]string // Maps username/prefix to password
+}
+
+// Valid checks if the provided prefix and password are valid
+func (s *StaticCredentialStore) Valid(prefix, password, userAddr string) bool {
+	storedPassword, exists := s.credentials[prefix]
+	return exists && storedPassword == password
+}
+
+func udpAssociateHandler(ctx context.Context, writer io.Writer, request *socks5.Request) error {
+	// Extract the suffix from the auth context
+	suffix, ok := request.AuthContext.Payload["suffix"]
+	if !ok {
+		return errors.New("no suffix found in auth context")
+	}
+
+	soaxPackage := viper.GetString("soax.package")
+	soaxPassword := viper.GetString("soax.password")
+	soaxAddress := viper.GetString("soax.address")
+
+	// Construct the upstream soax username by concatenating
+	// the soax package name with the suffix
+	soaxUsername := soaxPackage + suffix
+	log.Printf("Sending associate command with username: %s\n", soaxUsername)
+
+	streamEndpoint := transport.StreamDialerEndpoint{Dialer: &transport.TCPDialer{}, Address: soaxAddress}
+	client, err := csocks5.NewClient(&streamEndpoint)
+	if err != nil {
+		return err
+	}
+
+	err = client.SetCredentials([]byte(soaxUsername), []byte(soaxPassword))
+	if err != nil {
+		return err
+	}
+
+	client.EnablePacket(&transport.UDPDialer{})
+
+	conn, bindAddr, err := client.ConnectAndRequest(ctx, csocks5.CmdUDPAssociate, "0.0.0.0:0")
+	if err != nil {
+		return err
+	}
+
+	// Start a goroutine to close the connection after a timeout
+	go func() {
```

Comment:
Replace this timeout with a loop that reads from the client TCP connection and closes the target TCP connection and the client UDP connection when the client TCP connection is closed. The "loop" can be an io.Copy directly into io.Discard.